File: Problems/733_Flood_Fill.cpp
#include<vector>
#include<queue>
using namespace std;

// bfs
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int original_color = image[sr][sc];
        if(original_color==color) return image;
        queue<pair<int, int>> qu;
        vector<pair<int, int>> directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        qu.push({sr, sc});
        image[sr][sc] = color;
        while(!qu.empty()){
            pair<int, int> curr = qu.front();
            int r = curr.first;
            int c = curr.second;
            qu.pop();
            for(auto dir: directions){
                int dr = dir.first;
                int dc = dir.second;
                int nxt_r = r+dr;
                int nxt_c = c+dc;
                if(image.size()>nxt_r && nxt_r>=0 && image[0].size()>nxt_c && nxt_c>=0 and image[nxt_r][nxt_c]==original_color){
                    image[nxt_r][nxt_c] = color;
                    qu.push({nxt_r, nxt_c});
                }
            }
        }
        return image;
    }
};
// dfs
// class Solution {
// public:
//     vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
//         int original_color = image[sr][sc];
//         if(original_color==color) return image;
//         dfs(image, sr, sc, color, original_color);
//         return image;
//     }
//     void dfs(vector<vector<int>>& image, int sr, int sc, int color, int original_color){
//         if(image[sr][sc]==original_color){
//             image[sr][sc] = color;
//             if(sr-1>=0)                 dfs(image, sr-1, sc, color, original_color);
//             if(image.size()>sr+1)       dfs(image, sr+1, sc, color, original_color);
//             if(sc-1>=0)                 dfs(image, sr, sc-1, color, original_color);
//             if(image[0].size()>=sc+1)   dfs(image, sr, sc+1, color, original_color);
//         }
//     }
// };

File: Problems/733_Flood_Fill.py
from collections import deque
from typing import List



# dfs
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original_color = image[sr][sc]
        if original_color == color:
            return image
        m = len(image)
        n = len(image[0])
        def dfs(row, col):
            if image[row][col] == original_color:
                image[row][col] = color
                if row-1>=0:
                    dfs(row-1, col)
                if m>row+1:
                    dfs(row+1, col)
                if col-1>=0:
                    dfs(row, col-1)
                if n>col+1:
                    dfs(row, col+1)
        dfs(sr, sc)
        return image

# dfs but cleaner
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        originColor = image[sr][sc]
        if originColor == color:
            return image
        ROW = len(image)
        COL = len(image[0])
        directions = ((0, 1), (1, 0), (0, -1), (-1, 0))
        def traverse(r, c):
            if not(ROW>r>=0 and COL>c>=0 and image[r][c]==originColor):
                return
            image[r][c] = color
            for dr, dc in directions:
                traverse(r+dr, c+dc)
        traverse(sr, sc)
        return image

# using bfs and "deque" could make it faster
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original_color = image[sr][sc]
        if color == original_color:
            return image
        m = len(image)
        n = len(image[0])
        directions = ((1, 0), (0, 1), (-1, 0), (0, -1))
        image[sr][sc] = color
        queue = deque([[sr, sc]])
        while queue:
            x, y = queue.popleft()
            for dir in directions:
                dx, dy = dir
                nxt_x = x+dx
                nxt_y = y+dy
                if m>nxt_x>=0 and n>nxt_y>=0 and image[nxt_x][nxt_y] == original_color:
                    image[nxt_x][nxt_y] = color
                    queue.append([nxt_x, nxt_y])
        return image

# bfs using tuple
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        ROW = len(image)
        COL = len(image[0])
        originColor = image[sr][sc]
        if color == originColor:
            return image
        image[sr][sc] = color
        directions = ((0, 1), (1, 0), (0, -1), (-1, 0))
        qu = deque([(sr, sc)])
        
        while qu:
            r, c = qu.popleft()
            for dr, dc in directions:
                nr, nc = r+dr, c+dc
                if ROW>nr>=0 and COL>nc>=0 and image[nr][nc] == originColor:
                    image[nr][nc] = color
                    qu.append((nr, nc))
        return image

# bfs
# class Solution:
#     def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
#         if color == image[sr][sc]:
#             return image
#         m = len(image)
#         n = len(image[0])
#         directions = ((1, 0), (0, 1), (-1, 0), (0, -1))
#         original_color = image[sr][sc]
#         image[sr][sc] = color
#         curr = [[sr, sc]]
#         while curr:
#             nxt = []
#             for coord in curr:
#                 x, y = coord
#                 for dir in directions:
#                     dx, dy = dir
#                     nxt_x = x+dx
#                     nxt_y = y+dy
#                     if m>nxt_x>=0 and n>nxt_y>=0 and image[nxt_x][nxt_y] == original_color:
#                         image[nxt_x][nxt_y] = color
#                         nxt.append([nxt_x, nxt_y])
#                 curr = nxt
#         return image

if __name__ =='__main__':
    image = [[1,1,1],[1,1,0],[1,0,1]]
    sr = 1
    sc = 1
    color = 2
    sol = Solution()
    print(sol.floodFill(image, sr, sc, color))